src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -560,10 +549,24 @@ private void ProduceTagsSynchronously(TaggerContext<TTag> context)
                     // If we're on the foreground already, we can just update our internal state directly.
                     UpdateStateAndReportChanges(newTagTrees, bufferToChanges, newState, initialTags);
                 }
+                else if (initialTags)
+                {
+                    // If this is the initial set of tags, we fast-track a notification about whatever initial tags we
+                    // computed.  This way the UI is updated quickly for that initial set, and we don't have to wait a
+                    // potentially very long time as the foreground-thread-queue makes it way to our notification.
+                    this.ThreadingContext.JoinableTaskFactory.Run(async () =>
+                    {
+                        await this.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                        UpdateStateAndReportChanges(newTagTrees, bufferToChanges, newState, initialTags);
+                    });
```

Comment:
notes:

1. this only happens once.  so it's only N UI requests for our N taggers that actually end up running.
2. not doing this adds a very significant amount of waiting time for tagging results when a doc opens.  this is because 'RegisterNotification' uses the IForegroundNotificationService which was designed to *aggressively* back off to not impact the UI thread.

This approach here gets a good blaance.  For all but one tagger notificaiton we still aggressively back off, so we don't impact the UI thread.  However, we let at least one tagger result through initially so it does not seem as if things are stalled out when a file is opened.

Reply:
> not doing this adds a very significant amount of waiting time for tagging results when a doc opens

This should not be a true statement if the registration is occurring with a delay of 0 (which is the delay used for the next section). For a delay of 0, a backoff of at most 50ms will apply, and then the operation is invoked on the UI thread with the same priority as what you have here.

Reply:
so this is not what i observe.  please see the two attached gifs.  The first is for when we immediatley bounce to the UI thread here.  In this, the time delay is around 2.66s (measured by counting actual recorded capture frames) until the classification appears.

The second is removing this section and just using the next section (which has the 0 delay as you mention).  With this, it takes a full 10+ seconds to finally appear.  The results of direct notification are consistent.  THe results of using the FG notification queue are erratic.  I've had it as low as 7 seconds, but also higher as well.

Looking at teh code for it, i do not see what guarantees the behavior you're specifying.  Specifically, when you put it in with a delay of 0, it still goes after all existing work items that have a time that might be requested to fire prior to the current tick count (but which we haven't gotten to):

https://github.com/dotnet/roslyn/blob/1aeda2e94fb9371b96d4bfd94b074860064ec3d2/src/EditorFeatures/Core/Implementation/ForegroundNotification/ForegroundNotificationService.cs#L332-L338

This then does a InvokeBelowInputPriorityAsync once the BG task gets to it, which will then switch to the UI thread. However, when this UI thread starts processing items, it *aggressively backs off*.  Specifically:

https://github.com/dotnet/roslyn/blob/1aeda2e94fb9371b96d4bfd94b074860064ec3d2/src/EditorFeatures/Core/Implementation/ForegroundNotification/ForegroundNotificationService.cs#L184-L188

I'm assuming IsInputPending isn't firing.  However, if we just examine the second clause we can see an issue.  Say we get added to the end of N items in the queue that are set to run, and those items themselves each take more than 15ms to run (the value of DefaultTimeSliceInMS).  Then we effectively have the following:

Process an item.  Return out of this because we've crossed this threshold.  Wait for tasks to resolve back to BG.  BG thread then spins, see the next item.  It kicks off to the UI thread to process.  UI thread processes that item.   Repeat.

Effectively, this is really good at trying to space things out and perform work on this shared queue in order.  However, it doesn't actually totally guarantee that items taht were requested with a delay of 0 are next to run, or that they will all run before attempting to throttle back.

With the added section:

![immediate](https://user-images.githubusercontent.com/4564579/91486945-d6b6d700-e861-11ea-9aea-53495645b629.gif)

With 'FG notification service':
![foreground notification](https://user-images.githubusercontent.com/4564579/91486960-e0403f00-e861-11ea-9ee2-e2b5cf3025e6.gif)

Reply:
not sure what happened with teh BG capture for the first gif.  however, you can see how much faster the semantic classifications appear versus the second.